src/serde/ser.rs

```diff
@@ -0,0 +1,328 @@
+    }
+}
+
+pub(crate) struct StructVariantSerializer {
+    variant: &'static str,
+    fields: Vec<(Value, Value)>,
+}
+impl<'a> ser::SerializeStructVariant for StructVariantSerializer {
+    type Ok = Value;
+    type Error = Error;
+
+    fn serialize_field<T: ?Sized + Serialize>(
+        &mut self,
+        key: &'static str,
+        value: &T,
+    ) -> Result<(), Self::Error> {
+        let mut s = Serializer {};
+        self.fields
+            .push((key.serialize(&mut s)?, value.serialize(&mut s)?));
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        // Normalization (sorting) also happens inside sk_cbor, so this is
+        // technically redundant, but oasis_cbor (non-serde) does the same,
+        // so this makes the two compatible at the `Value` level.
+        let mut fields = self.fields;
+        fields.sort();
+
+        Ok(Value::Map(vec![(
+            self.variant.serialize(&mut Serializer {})?,
+            Value::Map(fields),
+        )]))
+    }
+}
+
+pub(crate) struct TupleVariantSerializer {
+    variant: &'static str,
+    fields: Vec<Value>,
+}
+impl<'a> ser::SerializeTupleVariant for TupleVariantSerializer {
+    type Ok = Value;
+    type Error = Error;
+
+    fn serialize_field<T: ?Sized + Serialize>(&mut self, value: &T) -> Result<(), Self::Error> {
+        self.fields.push(value.serialize(&mut Serializer {})?);
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(Value::Map(vec![(
+            self.variant.serialize(&mut Serializer {})?,
+            Value::Array(self.fields),
+        )]))
+    }
+}
+
+pub(crate) struct TupleStructSerializer {
+    fields: Vec<Value>,
+}
+impl ser::SerializeTupleStruct for TupleStructSerializer {
+    type Ok = Value;
+    type Error = Error;
+
+    fn serialize_field<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.fields.push(value.serialize(&mut Serializer {})?);
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(Value::Array(self.fields))
+    }
+}
+
+pub(crate) struct MapSerializer {
+    key: Option<Value>,
+    items: Vec<(Value, Value)>,
+}
+impl ser::SerializeMap for MapSerializer {
+    type Ok = Value;
+    type Error = Error;
+
+    fn serialize_key<T: ?Sized + Serialize>(&mut self, key: &T) -> Result<(), Self::Error> {
+        self.key = Some(key.serialize(&mut Serializer {})?);
+        Ok(())
+    }
+
+    fn serialize_value<T: ?Sized + Serialize>(&mut self, value: &T) -> Result<(), Self::Error> {
+        self.items.push((
+            self.key
+                .take()
+                .expect("serde tried to encode a map value without a key"),
+            value.serialize(&mut Serializer {})?,
+        ));
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        // Normalization (sorting) also happens inside sk_cbor, so this is
```

Comment:
Note that we currently only sort for struct (and variant struct) fields, not for regular maps.

Reply:
Removed sorting, added explanation.